specifications/cloud_assembly.md

```diff
@@ -0,0 +1,398 @@
+  * An error **MUST** be raised if the [PGP][RFC 4880] key has expired according to the signature timestamp.
+  * An error **MUST** be raised if the [PGP][RFC 4880] key is known to have been revoked. Deployment systems **MAY**
+    trust locally available information pertaining to the key's validity.
+* **SHOULD** allow the user to specify a list of trusted [PGP][RFC 4880] keys.
+
+## Annex
+### Examples of Drops for the AWS Cloud
+#### `@aws-cdk/aws-cloudformation.StackDrop`
+A [*CloudFormation* stack][CFN Stack].
+
+##### Properties
+Property     |Type                |Required|Description
+-------------|--------------------|:------:|-----------
+`template`   |`string`            |Required|The assembly-relative path to the *CloudFormation* template document.
+`parameters` |`Map<string,string>`|        |Parameters to be passed to the [stack][CFN Stack] upon deployment.
+`stackPolicy`|`string`            |        |The assembly-relative path to the [Stack Policy][CFN Stack Policy].
+
+##### Output Attributes
+Attribute |Type                |Description
+----------|--------------------|-----------
+`outputs` |`Map<string,string>`|Data returned by [*CloudFormation* Outputs][CFN Outputs] of the stack.
+`stackArn`|`string`            |The ARN of the [stack][CFN Stack].
+
+##### Example
+```json
+{
+  "type": "npm://@aws-cdk/aws-cloudformation.StackDrop",
+  "environment": "aws://000000000000/bermuda-triangle-1",
+  "properties": {
+    "template": "my-stack/template.yml",
+    "parameters": {
+      "bucketName": "${helperStack.bucketName}",
+      "objectKey": "${helperStack.objectKey}"
+    },
+    "stackPolicy": "my-stack/policy.json"
+  }
+}
+```
+
+[CFN Stack]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/stacks.html
+[CFN Stack Policy]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/protect-stack-resources.html
+[CFN Outputs]: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/outputs-section-structure.html
+
+#### `@aws-cdk/assets.FileDrop`
+A file that needs to be uploaded to an *S3* bucket.
+
+##### Properties
+Property    |Type    |Required|Description
+------------|--------|:------:|-----------
+`file`      |`string`|Required|The assembly-relative path to the file that will be uploaded.
+`bucketName`|`string`|Required|The name of the bucket where this file will be uploaded.
+`objectKey` |`string`|Required|The key at which to place the object in the bucket.
+
+##### Output Attributes
+Attribute   |Type    |Description
+------------|--------|-----------
+`bucketName`|`string`|The name of the bucket where the file was uploaded.
+`objectKey` |`string`|The key at which the file was uploaded in the bucket.
+
+##### Example
+```json
+{
+  "type": "npm://@aws-cdk/assets.FileDrop",
+  "environment": "aws://000000000000/bermuda-triangle-1",
+  "properties": {
+    "file": "assets/file.bin",
+    "bucket": "${helperStack.bucketName}",
```

Comment:
What if the `helperStack` is the bootstrap stack? Does a CDK app always add in the bootstrap stack into the assembly, or is the bootstrap stack supplied as a parameter to the assembly?

Reply:
So this is the case where you explicitly specify "everything" in your cloud assembly, but in case your "provisioning orchestrator" does this for you, then it wouldn't be part of the assembly. That's where the optional inputs I referred to in a previous reply come into play - the provisioning orchestrator wires it's own values in at deploy time, and they are available for reading from the asset `Drop`'s outputs.